### DISCUSSION TITLE
apache wsgi venv: libclntsh.so not found but on console it is

### DISCUSSION THREAD

**Discussion by schaechtelchen** [oracle/python-oracledb#160] [Q&A] [answered]
Hi there,

My application is called via apache wsgi and uses a venv.
I need to use thick client and have installed instantclient under /opt/oracle. 
But `oracledb.init_oracle_client()` throws the following error (with DPI_DEBUG_LEVEL = 64)

> [Fri Mar 24 12:23:36.572641 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "/var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/flask/app.py", line 1799, in dispatch_request
[Fri Mar 24 12:23:36.572645 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
[Fri Mar 24 12:23:36.572648 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "/var/www/ezdw.praxis.pikt/ezdw/ezdw.py", line 55, in suids_update
[Fri Mar 24 12:23:36.572653 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]     risdb = get_risdb()
[Fri Mar 24 12:23:36.572667 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "/var/www/ezdw.praxis.pikt/ezdw/db.py", line 27, in get_risdb
[Fri Mar 24 12:23:36.572670 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]     oracledb.init_oracle_client()
[Fri Mar 24 12:23:36.572673 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "src/oracledb/impl/thick/utils.pyx", line 455, in oracledb.thick_impl.init_oracle_client
[Fri Mar 24 12:23:36.572675 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "src/oracledb/impl/thick/utils.pyx", line 479, in oracledb.thick_impl.init_oracle_client
[Fri Mar 24 12:23:36.572678 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382]   File "src/oracledb/impl/thick/utils.pyx", line 400, in oracledb.thick_impl._raise_from_info
[Fri Mar 24 12:23:36.572681 2023] [wsgi:error] [pid 1933964] [remote 10.8.0.146:44382] oracledb.exceptions.DatabaseError: DPI-1047: Cannot locate a 64-bit Oracle Client library: "libclntsh.so: cannot open shared object file: No such file or directory". See https://python-oracledb.readthedocs.io/en//latest/user_guide/initialization.html#setting-the-oracle-client-library-directory for help
ODPI [1933973] 2023-03-24 12:23:36.556: ODPI-C 4.6.0
ODPI [1933973] 2023-03-24 12:23:36.556: debugging messages initialized at level 64
ODPI [1933973] 2023-03-24 12:23:36.556: Context Parameters:
ODPI [1933973] 2023-03-24 12:23:36.556: Environment Variables:
ODPI [1933973] 2023-03-24 12:23:36.556:     LD_LIBRARY_PATH => "/opt/oracle/instantclient_21_8/"
ODPI [1933973] 2023-03-24 12:23:36.556: check module directory
ODPI [1933973] 2023-03-24 12:23:36.556: module name is /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/thick_impl.cpython-39-x86_64-linux-gnu.so
ODPI [1933973] 2023-03-24 12:23:36.556: load in dir /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb
ODPI [1933973] 2023-03-24 12:23:36.556: load with name /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/libclntsh.so
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/libclntsh.so: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with OS search heuristics
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.19.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.19.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.18.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.18.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.12.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.12.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.11.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.11.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.20.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.20.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: load with name libclntsh.so.21.1
ODPI [1933973] 2023-03-24 12:23:36.556: load by OS failure: libclntsh.so.21.1: cannot open shared object file: No such file or directory
ODPI [1933973] 2023-03-24 12:23:36.556: check ORACLE_HOME

Testing with a small script called by root in the same venv the call works flawesly

`import oracledb`
`db_config = {...}`
`oracledb.init_oracle_client()`
`connection = oracledb.connect(**db_config)`
`print("Database version:", connection.version)`
`connection.close()`

>(venv) root@doku-1:/var/www/ezdw.praxis.pikt/ezdw# python goi.py 
Database version: 19.12.0.0.0

As of
>ODPI [1933973] 2023-03-24 12:23:36.556: LD_LIBRARY_PATH => "/opt/oracle/instantclient_21_8/"

oracledb knows about the library path but doesn't find the librarys?
What's my problem?

**Comment by schaechtelchen** [comment]
Trying to run the testing script as user www-data like apache does, the library is found too:

>(venv) root@doku-1:/var/www/ezdw.praxis.pikt/ezdw# sudo -u www-data bash goi.sh 
pwd.struct_passwd(pw_name='www-data', pw_passwd='x', pw_uid=33, pw_gid=33, pw_gecos='www-data', pw_dir='/var/www', pw_shell='/usr/sbin/nologin')
ODPI [1934869] 2023-03-24 13:38:31.952: ODPI-C 4.6.0
ODPI [1934869] 2023-03-24 13:38:31.952: debugging messages initialized at level 64
ODPI [1934869] 2023-03-24 13:38:31.952: Context Parameters:
ODPI [1934869] 2023-03-24 13:38:31.952: Environment Variables:
ODPI [1934869] 2023-03-24 13:38:31.952:     LD_LIBRARY_PATH => "/opt/oracle/instantclient_21_8/"
ODPI [1934869] 2023-03-24 13:38:31.952: check module directory
ODPI [1934869] 2023-03-24 13:38:31.952: module name is /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/thick_impl.cpython-39-x86_64-linux-gnu.so
ODPI [1934869] 2023-03-24 13:38:31.953: load in dir /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb
ODPI [1934869] 2023-03-24 13:38:31.953: load with name /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/libclntsh.so
ODPI [1934869] 2023-03-24 13:38:31.953: load by OS failure: /var/www/ezdw.praxis.pikt/venv/lib/python3.9/site-packages/oracledb/libclntsh.so: cannot open shared object file: No such file or directory
ODPI [1934869] 2023-03-24 13:38:31.953: load with OS search heuristics
ODPI [1934869] 2023-03-24 13:38:31.953: load with name libclntsh.so
ODPI [1934869] 2023-03-24 13:38:31.957: load by OS successful
ODPI [1934869] 2023-03-24 13:38:31.957: validating loaded library
Database version: 19.12.0.0.0

**Comment by cjbj** [comment, marked as answer]
Try setting the library search path using 'ldconfig' instead of LD_LIBRARY_PATH. Check the python-oracledb installation instructions. Passing the environment variable through web server configurations and environments is notoriously delicate.

**Reply by schaechtelchen** [reply]
So I did and it worked. But I'd like to now what is going on there. 
ODPI obviously knows about the 
>LD_LIBRARY_PATH => "/opt/oracle/instantclient_21_8/"

And while
>load with OS search heuristics

is done, no web server is involved at that moment, isn't it?

**Reply by cjbj** [reply]
You might be able to use strace to see if there is something blocking loading.
